accepted/2020/platform-exclusion/platform-exclusion.md

```diff
@@ -58,61 +88,169 @@ unsupported.
-      unsupported on other platforms.
+The semantics of these new attributes are as follows:
+
+* An API that doesn't have any of these attributes is considered supported by
+  all platforms.
+* If either `[SupportedOSPlatform]` or `[UnsupportedOSPlatform]` attributes are
+  present, we group all attributes by OS platform identifier:
+    - **Allow list**. If the lowest version for each OS platform is a
+      `[SupportedOSPlatform]` attribute, the API is considered to *only* be
+      supported by the listed platforms and unsupported by all other platforms.
+    - **Deny list**. If the lowest version for each OS platform is a
+      `[UnsupportedOSPlatform]` attribute, then the API is considered to *only*
+      be unsupported by the listed platforms and supported by all other
+      platforms.
+    - **Inconsistent list**. If for some platforms the lowest version attribute
+      is `[SupportedOSPlatform]` while for others it is
+      `[UnsupportedOSPlatform]`, the analyzer will produce a warning on the API
+      definition because the API is attributed inconsistently.
+* Both attributes can be instantiated without version numbers. This means the
+  version number is assumed to be `0.0`. This simplifies guard clauses, see
+  examples below for more details.
+* `[ObsoletedInOSPlatform]` continuous to require a version number.
+* `[ObsoletedInOSPlatform]` by itself doesn't imply support. However, it doesn't
+  make sense to apply `[ObsoletedInOSPlatform]` unless that platform is
+  supported.
+
+These semantics result in intuitive attribute applications with flexible rules.
+
+Let's consider the example where in .NET 5 we mark an API as being unsupported
+in `Windows`. This will look as follows:
+
+```C#
+[UnsupportedOSPlatform("windows")]
+public void DoesNotWorkOnWindows();
+```
+
+Now let's say that in .NET 6 we made the API work in Windows, but only on
+Windows `10.0.1903`. We'd update the API as follows:
+
+```C#
+[UnsupportedOSPlatform("windows")]
+[SupportedOSPlatform("windows10.0.1903")]
+public void DoesNotWorkOnWindows();
+```
+
+The presence of `SupportedOSPlatform` still makes no claim for other platforms
+because the lowest version for Windows says the API is unsupported for Windows.
+
+Similar to [platform-checks] this model still allows for OS vendors to obsolete
+and remove APIs:
+
+```C#
+[UnsupportedOSPlatform("windows")]
+[SupportedOSPlatform("windows10.0.1903")]
+[ObsoletedInOSPlatform("windows10.0.1909")]
+[SupportedOSPlatform("windows10.0.2004")]
+public void DoesNotWorkOnWindows();
+```
+
+This describes an API that is supported by all platforms, except for Windows.
+And on Windows, it's supported for Windows 10 1903 and 1909.
+
+Platform-specific APIs like the iOS and Android bindings are still expressible
+the same way:
+
+```C#
+[SupportedOSPlatform("ios12.0")]
+[ObsoletedInOSPlatform("ios13.0")]
+[UnsupportedOSPlatform("ios14.0")]
+[SupportedOSPlatform("ipados13.0")]
+public void OnlyWorksOniOS();
```

Comment:
Should we use a `params string[] platformNames`-based constructor for the attributes?  If so, we could then have the following:

```suggestion
[SupportedOSPlatform("ios12.0", "ipados13.0")]
[ObsoletedInOSPlatform("ios13.0")]
[UnsupportedOSPlatform("ios14.0")]
public void OnlyWorksOniOS();
```

With the iOS family of platforms, this might become a common need.

Reply:
Array might work if we know all supported platforms when creating the attribute. But for adding a new platform supported we will need to add a new element to the array? isn't it mean we will need update the existing attribute? By my understanding, we want to avoid updating existing attributes

Reply:
See [notes here](https://github.com/dotnet/runtime/issues/39989#issuecomment-665197220). We decided against `params.